File: src/main/java/com/zerone/core/SceneManager.java
package com.zerone.core;

public class SceneManager {

    private static Scene scene;

    public static final Event<Scene> onSceneLoaded;
    public static final Event<Scene> onSceneUnloaded;

    static {
        onSceneLoaded = new Event<>();
        onSceneUnloaded = new Event<>();
    }

    public static void loadScene(Scene newScene) {
        if (newScene == null)
            throw new IllegalArgumentException("Scene must not be null");

        if (scene != null) {
            scene.onPause();
            scene.onDestroy();

            onSceneUnloaded.invoke(scene);
        }

        newScene.onCreate();
        newScene.onResume();

        scene = newScene;

        onSceneLoaded.invoke(scene);
    }

    public static Scene getLoadedScene() {
        return scene;
    }

    static void resume() {
        scene.onResume();
    }

    static void pause() {
        scene.onPause();
    }
}

File: src/main/kotlin/com/zerone/core/SceneManager.kt
package com.zerone.core

object SceneManager {

    private var scene: Scene? = null

    val onSceneLoad: Event<Scene> = Event()
    val onSceneUnload: Event<Scene> = Event()

    fun loadScene(newScene: Scene) {
        scene?.let {
            it.onPause()
            it.onDestroy()

            onSceneUnload(it)
        }

        newScene.onCreate()
        newScene.onResume()

        scene = newScene

        onSceneLoad(newScene)
    }
}
